### PR TITLE
Provide a custom `nn.Module` class for better integration with metatensor data

### LINKED ISSUE #875: Compatibility of TensorMaps and TensorBlocks with buffers in PyTorch model?
Greetings  to my fellow metaverse peeps 👋 

There are several cases where it would be nice for the user to be able to save TensorMaps and TensorBlocks as buffers in PyTorch Models, which are then communicated through the state dicts of the models during checkpoint routines. Example within the context of MLIPs would be composition weights (baseline energies), which was something that I ran into recently with @frostedoyster.

AFAIK this is not yet possible, and hence one needs to come up with rather clunky workarounds to avoid such an issue: https://github.com/metatensor/metatrain/pull/522

It'd be nice if this issue could be discussed and resolved, thanks!

### PR DESCRIPTION
**Pull request by Luthaf**
This adds a `mts.learn.nn.Module` class that can be used instead of `torch.nn.Module` and provides integration between torch and metatensor data in two ways:

- calling `module.to()` will also move Labels, TensorBlock, TensorMap, as well as dict/list/tuple containing them to the correct dtype/device
    - in C++ mode we need a separate class to implement `to` since `torch::jit::Module` and `torch::nn::Module` are completely separate from `torch.nn.Module`
    - same when running TorchScript code from Python, we have to monkeypatch `torch.jit.RecursiveScriptModule`
- `state_dict()` and `load_state_dict` will properly save and load metatensor data
    - this is also true for float32 data/on device data. For now the code will convert to float64 and back below the hood, until #775 is fixed

TODO: 
- [x] document the format used in state_dict
- [x] rebase once #920 is merged

Fix #875 

# Contributor (creator of pull-request) checklist

 - [x] Tests updated (for new features and bugfixes)?
 - [x] Documentation updated (for new features)?
 - [x] Issue referenced (for PRs that solve an issue)?

# Reviewer checklist

 - [x] CHANGELOG updated with public API or any other important changes?

### CODE REVIEW

**Comment by Luthaf** [comment]
Still TBD: using this inside the different modules. One issue will be that until we update `load_atomistic_model` to use the custom C++ Module, we'll need to keep the code that moves data to dtype/device in `forward`.

File: metatensor-torch/tests/test-module.pt
Comment (jwa7) [inline metatensor-torch/tests/test-module.pt:1]: This looks like it's created by the `create-test-module.py` - should it still be here?

File: python/metatensor_learn/metatensor/learn/nn/_module.py [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+            some_changed = some_changed or changed
+            updated.append(updated_value)
+
+        if some_changed:
+            return tuple(updated), True
+
+    return value, False
+
+
+def _serialize_metatensor(value):
```
Comment (jwa7) [inline python/metatensor_learn/metatensor/learn/nn/_module.py:122]: How about `LabelsEntry`?

**Review by jwa7 — COMMENTED** [review]
Looks good on a first pass, but a couple of comments:

File: metatensor-torch/tests/test-module.pt
Comment (Luthaf) [reply metatensor-torch/tests/test-module.pt:1]: The issue is that we need an installation of metatensor-torch to create this file, and it is used to then test metatensor-torch itself … I'd like to dynamically generate this file, but doing so might make running the tests a lot harder. Hence why this file is committed to this repository.

File: python/metatensor_learn/metatensor/learn/nn/_module.py (hunk 1 above)
Comment (Luthaf) [reply python/metatensor_learn/metatensor/learn/nn/_module.py:122]: I'd leave it out for now, unless there is a pressing need for it. I'd rather implement #568 as a solution here.

Adding it would first require to define the serialization function for LabelsEntry in C++.

File: metatensor-torch/src/module.cpp [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+
+using namespace metatensor_torch;
+
+template <typename T>
+bool is_custom_class(torch::IValue ivalue) {
+    if (ivalue.isCustomClass()) {
+        // this is inspired by the code inside `torch::IValue.toCustomClass<T>()`
+        auto* expected_type = torch::getCustomClassType<torch::intrusive_ptr<T>>().get();
+        return ivalue.type().get() == expected_type;
```
[the remark this answers is not in the document]
Comment (HaoZeke) [reply metatensor-torch/src/module.cpp:18]: Also perhaps better to use `c10::` over `torch::` to be more self documenting (veryminor nit)

File: metatensor-torch/src/module.cpp [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+                    "as values are not supported"
+                );
+            }
+            return std::make_pair(updated, true);
+        }
+    } else if (ivalue.isList()) {
+        const auto& list = ivalue.toList();
+        auto updated = c10::impl::GenericList(list.elementType());
+        auto all_changed = true;
+        auto some_changed = false;
```
Comment (HaoZeke) [inline metatensor-torch/src/module.cpp:70]: Minor nit again, I'd prefer these to be set to falsy and updated to truthy in the loop, i.e. `false and true` instead of true and false.
